src/transforms/ppmt.jl

```diff
@@ -0,0 +1,171 @@
+
+struct Ppmt
+    table
+    tol::Real
+    perc::Real
+    trunc::Integer
+    sample::Integer
+end
+
+Ppmt(table) = Ppmt(table,1e-5,.9,10,10)
+
+function normal_score_transform(D, ϵ = 1e-5)
+    #Modifies the aleatory variables so they have standard normal cdf.
+    _, q = size(D)
+    Y = Matrix{Float64}(undef,size(D))
+    for i = 1:q #Computing over the q aleatory variables zᵢ
+        cdf = EmpiricalCDF(D[:,i])
+        sort!(cdf)
+        Y[:,i] = [quantile(Normal(), q) for q in cdf(D[:,i]).-ϵ]
+    end
+    return Y
+end
+
+function data_sphering(Y)
+    #Normalize the random variables (send them to the unitary sphere)
+    Σ = cov(Y,dims=1)
+    d, U = eigen(Σ)
+    D = Diagonal(d)
+    return Y*U*(D^(-.5))*U'
+end
+
+function r_transform(Z,α)
+    #Transforms a row of r.v. into a convex combination r.v. with values in [-1,1] #and standard normal distribution.
+    α_unit = (1/norm(α)) .* α .* (norm(α) != 1) + α .* (norm(α) == 1)
+    X = Z * α_unit
+    2 * cdf(Normal(), X) .- 1
+end
+
+function projection_index(Z,α,J)
+    #Generates the projection index I defined in the paper:
+    #Friedman, J. H. (1987). Exploratory Projection Pursuit. Journal of the American Statistical Association, 82(397), 249–266. https://doi.org/10.2307/2289161
+    R = r_transform(Z,α)
+    if J <= 0
+        return "Error: J must be non-negative"
+    elseif J == 1
+        return (3/2) * mean(R)^2
+    else
+        I = (3/2) * mean(R)^2
+        Pⱼ₋₂, Pⱼ₋₁ = ones(size(R)), R
+        for j = 2:J
+            Pⱼ₋₂, Pⱼ₋₁ = Pⱼ₋₁, (1/j) * ((2j-1) * R .* Pⱼ₋₁ - (j-1) * Pⱼ₋₂)
+            I += ((2j+1)/2) * (mean(Pⱼ₋₁))^2
+        end
+    end
+    return I
+end
+
+function max_principal(Z,J)
+    #Finds the canonical (positive) vector eᵢ that maximizes the projection index I
+    _, q = size(Z)
+    Iₘₐₓ, iₘₐₓ = 0, 0
+    for i = 1:q
+        eᵢ = Float64[1 * (j==i) for j=1:q]
+        I = projection_index(Z,eᵢ,J)
+        if I > Iₘₐₓ
+            Iₘₐₓ, iₘₐₓ = I, i
+        end
+    end
+    return Float64[1 *(iₘₐₓ==i) for i = 1:q]
+end
+
+function max_diagonal(Z,J)
+    #Given the maximizer principal direction, finds the diagonal direction that #maximizes I.
+    ᾱ = max_principal(Z,J)
+    l = length(ᾱ)
+    I₀ = projection_index(Z,ᾱ,J)
+    for i in 1:l
+        eᵢ = Float64[1*(j==i) for j=1:l]
+        f₊ = projection_index(Z, (1/√(2+2ᾱ[i]))*(ᾱ + eᵢ), J)
+        f₋ = projection_index(Z, (1/√(2-2ᾱ[i]))*(ᾱ - eᵢ), J) * (ᾱ[i] != 1)
+        f = (f₊>f₋)*f₊ + (f₊<=f₋)*f₋
+        s = (f₊>f₋)*1 + (f₊<=f₋)*(-1)
+        if f > projection_index(Z,ᾱ,J)
+            ᾱ = (1/√(2+2s*ᾱ[i]))*(ᾱ + s*eᵢ)
+        end
+    end
+    if projection_index(Z,ᾱ,J) >= I₀
+        return ᾱ
+    else
+        return max_principal(Z,J)
+    end
+end
+
+function optimizer(Z,J)
+    #Given the maximizer diagonal direction, finds a local maximum for I.
+    α₀ = max_diagonal(Z,J)
+    f(α) = -projection_index(Z,α,J)
+    op = optimize(f,α₀)
+    Optim.minimizer(op)
+end
```

Comment:
These three functions could be renamed to reflect the 3 optimization stages. Then, you could combine the three functions into a single function that produces the optimum direction for structure removal.